### PR TITLE
not a solution

### PR DESCRIPTION
**Pull request by d-beschastnyi**
pls i need help
i can't understand what i did wrong

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,214 +1,19 @@
...
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
+def battle(knightsConfig: dict) -> dict | str:
+    knights = []
 
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
+    # transforming input to Knight
+    for k in knightsConfig:
```
Comment (ArsenPidhoretskyi) [inline app/main.py:9]: k is not good name

File: app/main.py [hunk 2]
```diff
@@ -1,214 +1,19 @@
...
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
+    # LET THE BATTLE BEGIN
+    if len(knights) % 2 == 0:
+        for index in range(0, len(knights)-2, 2):
+            Battleground.fight(knights[index], knights[index+2])
+    else:
+        return "Not enough knights this year"
```
Comment (ArsenPidhoretskyi) [inline app/main.py:17]: Fighting must be by name, not by indexes

File: app/main.py [hunk 3]
```diff
@@ -1,214 +1,19 @@
...
-    return {
-        lancelot["name"]: lancelot["hp"],
-        arthur["name"]: arthur["hp"],
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
-    }
-
-
-print(battle(KNIGHTS))
+    return {knights[index].name: knights[index].hp for index in range(len(knights))}
```
Comment (ArsenPidhoretskyi) [inline app/main.py:19]: Iterate through values instead of index

File: app/modules/knights.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        self.name = knight["name"]
+        self.power = knight["power"]
+        self.hp = knight["hp"]
+        self.armour = knight["armour"]
+        self.weapon = knight["weapon"]
+        self.potion = knight["potion"]
+        self.protection = 0
+
+        # add "protection" stats from "armor" items
+        for a in self.armour:
```
Comment (ArsenPidhoretskyi) [inline app/modules/knights.py:12]: a is not good name

File: app/modules/knights.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            self.protection += a["protection"]
+
+        # add "power" stats from "weapon" items
+        for w in self.weapon:
+            self.power += w[self.weapon]
```
Comment (ArsenPidhoretskyi) [inline app/modules/knights.py:17]: It is just one item, not list of items
self.weapon - dict. Just do:
self.power += self.weapon["power"]

File: app/modules/knights.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        # add all stats from "potion" items
+        if self.potion is not None:
+            if "power" in self.potion["effect"]:
+                self.power += self.potion["effect"]["power"]
+
+            if "protection" in self.potion["effect"]:
+                self.protection += self.potion["effect"]["protection"]
+
+            if "hp" in self.potion["effect"]:
+                self.hp += self.potion["effect"]["hp"]
```
Comment (ArsenPidhoretskyi) [inline app/modules/knights.py:28]: Use forloop with getattr and setattr

File: app/main.py [hunk 7]
```diff
@@ -1,214 +1,19 @@
...
-        }
-    }
-}
+from app.modules.battle import Battleground
+from app.modules.knights import Knight
 
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
+def battle(knightsConfig: dict) -> dict | str:
```
Comment (ArsenPidhoretskyi) [inline app/main.py:5]:
```suggestion
def battle(knights_config: dict) -> dict | str:
```

**Review by ArsenPidhoretskyi — CHANGES_REQUESTED** [review]
Please fix flake8

**Review by vasylhnatiuk — COMMENTED** [review]
fix tests
image

**Review by vasylhnatiuk — CHANGES_REQUESTED** [review]
fix tests

File: app/main.py [hunk 8]
```diff
@@ -14,10 +14,19 @@ def fight(knight_1: Knight, knight_2: Knight) -> None:
...
     # LET THE BATTLE BEGIN
     fight(knights["lancelot"], knights["mordred"])
     fight(knights["arthur"], knights["red_knight"])
 
-    return {knight.name: knight.hp for knight in knights}
+    return {
+        "Lancelot": knights["lancelot"].hp,
+        "Arthur": knights["arthur"].hp,
+        "Mordred": knights["mordred"].hp,
+        "Red Knight": knights["red_knight"].hp
```
Comment (vasylhnatiuk) [inline app/main.py:31]: the better variant will be 

```
def battle(knights_config: dict):
    knights = {knight_name: Knight(knight_dict) for knight_name, knight_dict in knights_config.items()}

    # LET THE BATTLE BEGIN
    fight(knights["lancelot"], knights["mordred"])
    fight(knights["arthur"], knights["red_knight"])

    return {knight.name: knight.hp for knight in knights.values()}
```

File: app/main.py [hunk 9]
```diff
@@ -14,10 +14,19 @@ def fight(knight_1: Knight, knight_2: Knight) -> None:
...
+    }
```
Comment (vasylhnatiuk) [inline app/main.py]: and add newline

**Review by vasylhnatiuk — CHANGES_REQUESTED** [review]

**Review by ArsenPidhoretskyi — APPROVED** [review]
Well done!

**Pull request closed without merging**